File: GameLibrary/src/cc/lib/utils/StopWatch.java
package cc.lib.utils;

import java.io.IOException;

import cc.lib.reflector.Omit;
import cc.lib.reflector.RBufferedReader;
import cc.lib.reflector.Reflector;

public class StopWatch extends Reflector<StopWatch> {

    static {
        addAllFields(StopWatch.class);
    }

    @Omit
    protected long startTime = 0;
    @Omit
    protected long pauseTime = 0;
    @Omit
    protected long deltaTime = 0;
    @Omit
    protected long lastCaptureTime = 0;

    protected long curTime = 0;
    protected boolean started = false;
    
    /**
     * Start the stopwatch.  MUST be the first call
     */
    public void start() {
        curTime = 0;
        startTime = getClockMiliseconds();
        started = true;
        unpause();
    }

    public void stop() {
        started = false;
    }
    
    public boolean isPaused() {
        return pauseTime > 0;
    }

    public boolean isStarted() {
        return started;
    }
    
    /**
     * Pause the stop watch.  getTime/getDeltaTime will not advance until unpause called.
     */
    public void pause() {
        if (!isPaused()) {
            pauseTime = getClockMiliseconds();
        }
    }
    
    /**
     * Capture the current time and delta time.  Must be called before calling getTme, getDeltaTiime
     */
    public void capture() {
        if (started && pauseTime == 0) {
            long t = this.getClockMiliseconds();
            curTime += t - startTime;
            startTime = t;
            this.deltaTime = curTime - lastCaptureTime;
            this.lastCaptureTime = curTime;
        }
    }
    
    /**
     * Resume the stop watch if paused
     */
    public void unpause() {
        if (pauseTime > 0) {
            startTime = getClockMiliseconds();
            pauseTime = 0;
        }
    }
    
    /**
     * Get the time as of last call to capture()
     * @return
     */
    public long getTime() {
        if (started)
            return curTime;
        return 0;
    }
    
    /**
     * Get the delta time as of last call to capture()
     * @return
     */
    public long getDeltaTime() {
        if (started)
            return this.deltaTime;
        return 0;
    }
    
    /**
     * Override this to use a different clock mechanism if desired.
     * Default uses System.currentTimeMillis()
     * @return
     */
    protected long getClockMiliseconds() {
        return System.currentTimeMillis();
    }

    @Override
    protected void deserialize(RBufferedReader input) throws IOException {
        super.deserialize(input);
        if (started) {
            pauseTime = startTime = getClockMiliseconds();
        }
    }
}

File: GameLibrary2/src/cc/lib/utils/StopWatch.kt
package cc.lib.utils

import cc.lib.ksp.mirror.Mirror
import cc.lib.ksp.mirror.Mirrored
import com.google.gson.stream.JsonReader

@Mirror
interface IStopWatch : Mirrored {
	var curTime: Long
	val isStarted: Boolean
}

open class StopWatch(
	startTime: Long = 0,
	pauseTime: Long = 0,
	curTime: Long = 0,
	deltaTime: Long = 0,
	lastCaptureTime: Long = 0
) : StopWatchImpl() {

	var startTime: Long = startTime
		private set
	var pauseTime: Long = pauseTime
		private set
	var deltaTime: Long = deltaTime
		get() = if (isStarted) deltaTime else 0
		private set
	var lastCaptureTime: Long = lastCaptureTime
		private set

	init {
		this.curTime = curTime
	}

	/**
	 * Start the stopwatch.  MUST be the first call
	 */
	fun start() {
		curTime = 0
		startTime = clockMiliseconds
		isStarted = true
		unpause()
	}

	fun stop() {
		isStarted = false
	}

	val isPaused: Boolean
		get() = pauseTime > 0

	/**
	 * Pause the stop watch.  getTime/getDeltaTime will not advance until unpause called.
	 */
	fun pause() {
		if (!isPaused) {
			pauseTime = clockMiliseconds
		}
	}

	/**
	 * Capture the current time and delta time.  Must be called before calling getTme, getDeltaTiime
	 */
	fun capture() {
		if (isStarted && pauseTime == 0L) {
			val t = clockMiliseconds
			curTime += t - startTime
			startTime = t
			deltaTime = curTime - lastCaptureTime
			lastCaptureTime = curTime
		}
	}

	/**
	 * Resume the stop watch if paused
	 */
	fun unpause() {
		if (pauseTime > 0) {
			startTime = clockMiliseconds
			pauseTime = 0
		}
	}

	val time: Long
		/**
		 * Get the time as of last call to capture()
		 * @return
		 */
		get() = if (isStarted) curTime else 0

	protected open val clockMiliseconds: Long
		/**
		 * Override this to use a different clock mechanism if desired.
		 * Default uses System.currentTimeMillis()
		 * @return
		 */
		protected get() = System.currentTimeMillis()


	override fun fromGson(reader: JsonReader, __name: String) {
		super.fromGson(reader, __name)
		if (isStarted) {
			startTime = clockMiliseconds
			pauseTime = startTime
		}
	}

}
